### PR TITLE
attach secret to service account

### PR DESCRIPTION
**Pull request by deads2k**
Adds a command to attach secrets to a service account.  

```
oc secrets add default my-secret my-pull-secret-for-builds
oc secrets add default my-personal-pull-secret --for=pull
```

Builds on https://github.com/openshift/origin/pull/2866, but I don't want to hold up that pull.  Only the last commit is interesting.

@liggitt @smarterclayton @mfojtik

### CODE REVIEW

**Comment by deads2k** [comment]
updated to force users to qualify their names

**Comment by mfojtik** [comment]
@deads2k how about pushSecret?

**Comment by mfojtik** [comment]
@deads2k @liggitt also how about SourceSecret? This is currently also not part of the Service Account and referenced in the BuildConfig...

**Comment by liggitt** [comment]
@mfotjik service accounts have two lists of secrets: pull secrets for the kubelet to use and mountable secrets for containers to use. We're not going to have a command for every conceivable use of a mountable secret (push, source pull, database password, etc, etc)

**Comment by mfojtik** [comment]
@liggitt so you are saying that if I go ahead and create my secret foo with my private ssh key and then plug it into BuildConfig as SourceSecret, will it work or refused by admission controller? Similar question for PushSecret.

**Comment by deads2k** [comment]
> @liggitt so you are saying that if I go ahead and create my secret foo with my private ssh key and then plug it into BuildConfig as SourceSecret, will it work or refused by admission controller? Similar question for PushSecret

You'll need to add your secret reference to the BuildConfig **and** add it to the list of mountable secrets in the service account.

**Comment by mfojtik** [comment]
@deads2k right. I thought we are going to simplify the process for the users here. I think having private GIT repositories is pretty common use-case and having users to run `oc edit` sounds like a 'too much work' for me. I was hoping that we can have single command you run to achieve this.

**Comment by deads2k** [comment]
> @deads2k right. I thought we are going to simplify the process for the users here. I think having private GIT repositories is pretty common use-case and having users to run oc edit sounds like a 'too much work' for me. I was hoping that we can have single command you run to achieve this.

This command avoids the edit.  For a well behaved secret, you ought to be able to do something like 
```
oc secrets new my-push-secret ~/.dockercfg | xargs oc secrets add serviceaccount/builder
```

**Comment by mfojtik** [comment]
@deads2k thanks! that was the command I was looking for to add into documentation. For the SourceSecret I guess you have to use the serviceaccount/default, correct?

**Comment by liggitt** [comment]
the build pod pulls the source, right? so use the builder service account

**Comment by mfojtik** [comment]
@liggitt yes. thx!

**Comment by mfojtik** [comment]
@liggitt @deads2k also the source secret key contains '_' character :( (id_rsa)

**Comment by deads2k** [comment]
> @liggitt @deads2k also the source secret key contains '_' character :( (id_rsa)

Does it have to?  That is going to be a problem.  Kube enforces no '_'s.

**Comment by liggitt** [comment]
there's agreement to relax that upstream... we want to carry that?

**Comment by mfojtik** [comment]
@liggitt we we can, I would be happy to carry that path and fix upstream later... it will make our docs much nicer... we will have to fix the builders however.

**Comment by deads2k** [comment]
> there's agreement to relax that upstream... we want to carry that?

I'd rather not start persisting API objects that can be created by openshift, but are invalid for kube.

**Comment by liggitt** [comment]
yeah

**Comment by mfojtik** [comment]
agree... lets fix docs for now and come back to this post-3.0

**Comment by deads2k** [comment]
rebased

File: pkg/cmd/cli/secrets/addsecret_to_obj.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+	Mapper          meta.RESTMapper
+	Typer           runtime.ObjectTyper
+	ClientMapper    resource.ClientMapper
+	ClientInterface client.Interface
+
+	Out io.Writer
+}
+
+// NewCmdCreateDockerConfigSecret creates a command object for making a dockercfg secret
```
Comment (derekwaynecarr) [inline pkg/cmd/cli/secrets/addsecret_to_obj.go]: Bad doc.

File: pkg/cmd/cli/secrets/addsecret_to_obj.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+	cmdutil "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	AttachSecretRecommendedName = "add"
```
Comment (derekwaynecarr) [inline pkg/cmd/cli/secrets/addsecret_to_obj.go]: Can you update the constants to reflect the name?

After reviewing this, I forgot that the command was 'add' and not 'attach'.  I almost commented that I don't like the verb 'attach'.

File: pkg/cmd/cli/secrets/addsecret_to_obj.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+)
+
+type SecretType string
+
+var (
+	PullType  SecretType = "pull"
+	MountType SecretType = "mount"
+)
+
+type AttachSecretOptions struct {
```
Comment (derekwaynecarr) [inline pkg/cmd/cli/secrets/addsecret_to_obj.go]: AddSecretToServiceAccountOptions ?

**Comment by derekwaynecarr** [comment]
Nits on code names not matching user experience names.  Prefer we commit initially with the Add verb in code and not the Attach verb in code.  On a side note, do we need a corresponding undo command to remove a secret from a service account?

File: pkg/cmd/cli/secrets/addsecret_to_obj.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/cli/secrets/addsecret_to_obj.go]: > AddSecretToServiceAccountOptions ?

@smarterclayton wants to use the same command to handle adds to other target types.

**Comment by deads2k** [comment]
comments addressed.

**Comment by deads2k** [comment]
> On a side note, do we need a corresponding undo command to remove a secret from a service account?

`osc edit` in that direction is a lot easier because you know the name of the thing you're removing.  That makes it pretty easy to find the line in the json to delete, so I'm not so worried about it.

**Comment by deads2k** [comment]
Also, I'm pretty sure that our users do it perfectly the first time.

**Comment by deads2k** [comment]
@smarterclayton approval?

**Comment by derekwaynecarr** [comment]
>> Also, I'm pretty sure that our users do it perfectly the first time.

:laughing:

**Comment by smarterclayton** [comment]
Squash and approved

**Comment by deads2k** [comment]
[merge]

**Comment by liggitt** [comment]
```
++ Building go targets for linux/amd64: cmd/openshift
# github.com/openshift/origin/pkg/cmd/cli/secrets
_output/local/go/src/github.com/openshift/origin/pkg/cmd/cli/secrets/subcommand.go:34: undefined: NewCmdAttachSecret
_output/local/go/src/github.com/openshift/origin/pkg/cmd/cli/secrets/subcommand.go:34: undefined: AttachSecretRecommendedName
```

**Comment by deads2k** [comment]
re[merge]

**Comment by deads2k** [comment]
builder error: `An error occurred while installing mysql2 (0.3.16), and Bundler cannot continue.`

@bparees Known problem?

**Comment by deads2k** [comment]
@bparees I can reproduce on master.

**Comment by deads2k** [comment]
re[merge]

**Comment by deads2k** [comment]
build problem re[merge]

**Pull request merged**